File: examples/typescript/05-custom-sandboxed-tool/tests/tools.test.ts
/**
 * Unit tests for the two custom tools shipped by example 05.
 *
 * They are driven directly over an in-memory run store (no model, no Ollama),
 * proving each rule the example demonstrates:
 *   - `remember` stores under the `fact:` prefix, keyed by SessionId.
 *   - `recall` returns the stored value string.
 *   - a `recall` miss is a recoverable error with the exact message.
 *   - missing / wrong-typed args are recoverable errors.
 *   - a run-store failure is a recoverable error.
 *   - the read-only / idempotent annotations are correct.
 */

import {
  SessionId,
  storage,
  toolRegistry,
  type SandboxProvider,
  type SandboxViolation,
  type ToolCall,
  type ToolOutput,
} from "@spore/core";
import { describe, expect, it } from "vitest";

import { FACT_PREFIX, RememberTool } from "../src/tools/remember.js";
import { RecallTool } from "../src/tools/recall.js";

type RunStore = storage.RunStore;
type JsonValue = storage.JsonValue;

const { ToolContext } = toolRegistry;
const { InMemoryStorageProvider } = storage;

/** These tools never touch the filesystem — a permissive sandbox is plenty. */
class AllowAllSandbox implements SandboxProvider {
  async validate(_call: ToolCall): Promise<SandboxViolation | null> {
    return null;
  }
}

/** A RunStore that always fails — proves storage errors map to recoverable errors. */
class FailingRunStore implements RunStore {
  async get(
    _sessionId: SessionId,
    _key: string,
  ): Promise<JsonValue | undefined> {
    throw new Error("boom");
  }
  async put(
    _sessionId: SessionId,
    _key: string,
    _value: JsonValue,
  ): Promise<void> {
    throw new Error("boom");
  }
  async delete(_sessionId: SessionId, _key: string): Promise<void> {}
  async listKeys(_sessionId: SessionId): Promise<string[]> {
    return [];
  }
}

function ctxWith(
  runStore: RunStore,
  session = "test-session",
): toolRegistry.ToolContext {
  return new ToolContext(
    SessionId.of(session),
    runStore,
    new InMemoryStorageProvider(),
  );
}

function inMemoryCtx(): toolRegistry.ToolContext {
  return ctxWith(new InMemoryStorageProvider());
}

function rememberCall(input: unknown): ToolCall {
  return { id: "c1", name: RememberTool.NAME, input };
}

function recallCall(input: unknown): ToolCall {
  return { id: "c2", name: RecallTool.NAME, input };
}

function expectSuccess(out: ToolOutput): string {
  if (out.kind !== "success") {
    throw new Error(`expected success, got ${JSON.stringify(out)}`);
  }
  return out.content;
}

const sb = new AllowAllSandbox();

describe("RememberTool", () => {
  it("stores the value under the fact: prefix, keyed by SessionId", async () => {
    const ctx = ctxWith(new InMemoryStorageProvider(), "sess-a");
    const out = await new RememberTool().execute(
      rememberCall({ key: "habitat", value: "coastal ocean waters" }),
      sb,
      ctx,
    );
    expect(expectSuccess(out)).toBe("remembered habitat");

    // The blob lives under fact:habitat in the run store for this session.
    const stored = await ctx.runStore.get(
      SessionId.of("sess-a"),
      `${FACT_PREFIX}habitat`,
    );
    expect(stored).toBe("coastal ocean waters");
    // And NOT under the bare key.
    expect(await ctx.runStore.get(SessionId.of("sess-a"), "habitat")).toBe(
      undefined,
    );
  });

  it("missing 'key' is a recoverable error", async () => {
    const out = await new RememberTool().execute(
      rememberCall({ value: "x" }),
      sb,
      inMemoryCtx(),
    );
    expect(out.kind).toBe("error");
    if (out.kind === "error") expect(out.recoverable).toBe(true);
  });

  it("non-string 'value' is a recoverable error", async () => {
    const out = await new RememberTool().execute(
      rememberCall({ key: "k", value: 42 }),
      sb,
      inMemoryCtx(),
    );
    expect(out.kind).toBe("error");
    if (out.kind === "error") expect(out.recoverable).toBe(true);
  });

  it("a run-store failure is a recoverable error", async () => {
    const out = await new RememberTool().execute(
      rememberCall({ key: "k", value: "v" }),
      sb,
      ctxWith(new FailingRunStore()),
    );
    expect(out.kind).toBe("error");
    if (out.kind === "error") expect(out.recoverable).toBe(true);
  });

  it("is not read_only / destructive / idempotent", () => {
    const a = RememberTool.schema().annotations;
    expect(a.read_only).toBe(false);
    expect(a.destructive).toBe(false);
    expect(a.idempotent).toBe(false);
  });
});

describe("RecallTool", () => {
  it("returns the value a prior remember stored", async () => {
    const ctx = inMemoryCtx();
    await new RememberTool().execute(
      rememberCall({ key: "diet", value: "crabs and small fish" }),
      sb,
      ctx,
    );
    const out = await new RecallTool().execute(
      recallCall({ key: "diet" }),
      sb,
      ctx,
    );
    expect(expectSuccess(out)).toBe("crabs and small fish");
  });

  it("a miss is a recoverable error with the exact message", async () => {
    const out = await new RecallTool().execute(
      recallCall({ key: "nope" }),
      sb,
      inMemoryCtx(),
    );
    expect(out.kind).toBe("error");
    if (out.kind === "error") {
      expect(out.recoverable).toBe(true);
      expect(out.message).toBe("no fact stored under 'nope'");
    }
  });

  it("missing 'key' is a recoverable error", async () => {
    const out = await new RecallTool().execute(
      recallCall({}),
      sb,
      inMemoryCtx(),
    );
    expect(out.kind).toBe("error");
    if (out.kind === "error") expect(out.recoverable).toBe(true);
  });

  it("a run-store failure is a recoverable error", async () => {
    const out = await new RecallTool().execute(
      recallCall({ key: "k" }),
      sb,
      ctxWith(new FailingRunStore()),
    );
    expect(out.kind).toBe("error");
    if (out.kind === "error") expect(out.recoverable).toBe(true);
  });

  it("read does not write: a recall never persists anything", async () => {
    const ctx = inMemoryCtx();
    await new RecallTool().execute(recallCall({ key: "k" }), sb, ctx);
    expect(await ctx.runStore.listKeys(ctx.sessionId)).toEqual([]);
  });

  it("is read_only + idempotent (and not destructive)", () => {
    const a = RecallTool.schema().annotations;
    expect(a.read_only).toBe(true);
    expect(a.idempotent).toBe(true);
    expect(a.destructive).toBe(false);
  });
});

File: examples/go/05-custom-sandboxed-tool/tools/tools_test.go
package tools

import (
	"context"
	"encoding/json"
	"testing"

	sporecore "github.com/squirrelsoft-dev/spore-core/go/spore-core"
)

const testSession = sporecore.SessionID("sess-test")

// newCtx returns a ToolContext backed by a fresh in-memory run store — the same
// store the builder defaults to when .Tool() tools are present.
func newCtx() (*sporecore.ToolContext, sporecore.ToolRunStore) {
	store := sporecore.NewInMemoryToolRunStore()
	return sporecore.NewToolContext(testSession, store, nil), store
}

func call(name string, input string) sporecore.ToolCall {
	return sporecore.ToolCall{ID: "c1", Name: name, Input: json.RawMessage(input)}
}

// remember stores the value under the fact: prefix.
func TestRememberStoresUnderFactPrefix(t *testing.T) {
	ctx, store := newCtx()
	out := RememberTool{}.Execute(context.Background(), call("remember", `{"key":"habitat","value":"coastal waters"}`), nil, ctx)

	if out.Kind != sporecore.ToolOutputSuccess {
		t.Fatalf("Kind = %v, want success (message=%q)", out.Kind, out.Message)
	}
	if out.Content != "remembered habitat" {
		t.Errorf("Content = %q, want %q", out.Content, "remembered habitat")
	}

	// The raw store must carry the value under the namespaced key, JSON-encoded.
	raw, found, err := store.Get(context.Background(), testSession, FactPrefix+"habitat")
	if err != nil || !found {
		t.Fatalf("store.Get(fact:habitat) found=%v err=%v, want found", found, err)
	}
	var got string
	if err := json.Unmarshal(raw, &got); err != nil {
		t.Fatalf("stored value not a JSON string: %v", err)
	}
	if got != "coastal waters" {
		t.Errorf("stored value = %q, want %q", got, "coastal waters")
	}
	// And it must NOT be stored under the bare key.
	if _, found, _ := store.Get(context.Background(), testSession, "habitat"); found {
		t.Error("value leaked into the un-prefixed key 'habitat'")
	}
}

// recall returns the value a prior remember stored.
func TestRecallReturnsStoredValue(t *testing.T) {
	ctx, _ := newCtx()
	if out := (RememberTool{}).Execute(context.Background(), call("remember", `{"key":"diet","value":"crabs and shrimp"}`), nil, ctx); out.Kind != sporecore.ToolOutputSuccess {
		t.Fatalf("setup remember failed: %q", out.Message)
	}

	out := RecallTool{}.Execute(context.Background(), call("recall", `{"key":"diet"}`), nil, ctx)
	if out.Kind != sporecore.ToolOutputSuccess {
		t.Fatalf("Kind = %v, want success (message=%q)", out.Kind, out.Message)
	}
	if out.Content != "crabs and shrimp" {
		t.Errorf("Content = %q, want %q", out.Content, "crabs and shrimp")
	}
}

// recall on an unknown key is a recoverable error with the spec'd message.
func TestRecallMissingKeyIsRecoverableError(t *testing.T) {
	ctx, _ := newCtx()
	out := RecallTool{}.Execute(context.Background(), call("recall", `{"key":"nope"}`), nil, ctx)

	if out.Kind != sporecore.ToolOutputError {
		t.Fatalf("Kind = %v, want error", out.Kind)
	}
	if !out.Recoverable {
		t.Error("missing-key error must be recoverable")
	}
	if want := "no fact stored under 'nope'"; out.Message != want {
		t.Errorf("Message = %q, want %q", out.Message, want)
	}
}

// Bad arguments are recoverable errors, not panics or fatal errors.
func TestArgErrors(t *testing.T) {
	ctx, _ := newCtx()
	cases := []struct {
		name string
		tool sporecore.Tool
		args string
	}{
		{"remember missing key", RememberTool{}, `{"value":"v"}`},
		{"remember missing value", RememberTool{}, `{"key":"k"}`},
		{"remember non-string key", RememberTool{}, `{"key":123,"value":"v"}`},
		{"remember non-string value", RememberTool{}, `{"key":"k","value":123}`},
		{"remember empty input", RememberTool{}, ``},
		{"recall missing key", RecallTool{}, `{}`},
		{"recall non-string key", RecallTool{}, `{"key":123}`},
		{"recall empty input", RecallTool{}, ``},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			out := tc.tool.Execute(context.Background(), call(tc.tool.Name(), tc.args), nil, ctx)
			if out.Kind != sporecore.ToolOutputError {
				t.Fatalf("Kind = %v, want error", out.Kind)
			}
			if !out.Recoverable {
				t.Error("argument error must be recoverable")
			}
		})
	}
}

// Schema names must match Name(), and the read-only/idempotent annotations must
// reflect each tool's contract: remember mutates (not read-only), recall reads.
func TestSchemaAnnotations(t *testing.T) {
	rem := RememberTool{}
	if rem.Schema().Name != rem.Name() {
		t.Errorf("remember schema name %q != Name() %q", rem.Schema().Name, rem.Name())
	}
	if rem.Schema().Annotations.ReadOnly {
		t.Error("remember must NOT be read-only — it mutates shared state")
	}

	rec := RecallTool{}
	if rec.Schema().Name != rec.Name() {
		t.Errorf("recall schema name %q != Name() %q", rec.Schema().Name, rec.Name())
	}
	if !rec.Schema().Annotations.ReadOnly {
		t.Error("recall must be read-only")
	}
	if !rec.Schema().Annotations.Idempotent {
		t.Error("recall must be idempotent")
	}

	// Schemas must be valid JSON objects with the required fields.
	for _, s := range []sporecore.RegistryToolSchema{rem.Schema(), rec.Schema()} {
		var probe map[string]any
		if err := json.Unmarshal(s.Parameters, &probe); err != nil {
			t.Errorf("%s parameters are not valid JSON: %v", s.Name, err)
		}
	}
}

// A store failure surfaces as a recoverable error (not a panic). Drive it with a
// run store whose Put always fails.
func TestRememberStoreFailureIsRecoverable(t *testing.T) {
	ctx := sporecore.NewToolContext(testSession, failingStore{}, nil)
	out := RememberTool{}.Execute(context.Background(), call("remember", `{"key":"k","value":"v"}`), nil, ctx)
	if out.Kind != sporecore.ToolOutputError || !out.Recoverable {
		t.Fatalf("got Kind=%v Recoverable=%v, want recoverable error", out.Kind, out.Recoverable)
	}
}

type failingStore struct{}

func (failingStore) Get(context.Context, sporecore.SessionID, string) (json.RawMessage, bool, error) {
	return nil, false, errBoom
}
func (failingStore) Put(context.Context, sporecore.SessionID, string, json.RawMessage) error {
	return errBoom
}

var errBoom = errString("boom")

type errString string

func (e errString) Error() string { return string(e) }
